data/txHandler.go

```diff
@@ -319,7 +314,98 @@ func (handler *TxHandler) processDecoded(unverifiedTxGroup []transactions.Signed
 		logging.Base().Warnf("unable to pin transaction: %v", err)
 	}
 
-	return network.OutgoingMessage{}, false
+	return false
+}
+
+// filterAlreadyCommitted scan the list of signed transaction groups, and filter out the ones that have already been included,
+// or that should not be added to the transaction pool.
+// the resulting slice is using the *same* underlying array as the input slice, and the caller must ensure that this would not
+// cause issue on the caller side. The nonDuplicatedFilteredGroups describe whether any of the removed transacation groups was
+// removed for a reason *other* than being duplicate ( for instance, malformed transaction )
+func (handler *TxHandler) filterAlreadyCommitted(unverifiedTxGroups []transactions.SignedTxGroup) (filteredGroups []transactions.SignedTxGroup, nonDuplicatedFilteredGroups bool) {
+	remainedTxnsGroupOffset := 0
+	for idx, utxng := range unverifiedTxGroups {
+		err := handler.txPool.Test(utxng.Transactions)
+		switch err.(type) {
+		case nil:
+			// no error was generated.
+		case *ledgercore.TransactionInLedgerError:
+			// this is a duplicate transaction group.
+			continue
+		default:
+			// some non-duplicate error was reported on this group.
+			nonDuplicatedFilteredGroups = true
+			continue
+		}
+
+		if remainedTxnsGroupOffset != idx {
+			unverifiedTxGroups[remainedTxnsGroupOffset] = utxng
+		}
+		remainedTxnsGroupOffset++
+	}
+	return unverifiedTxGroups[:remainedTxnsGroupOffset], nonDuplicatedFilteredGroups
+}
+
+// processDecodedArray receives a slice of transaction groups and attempt to add them to the transaction pool.
+// The processDecodedArray returns whether the node should be disconnecting from the source of these transactions ( in case a malicious transaction is found )
+// as well as whether all the provided transactions were included in the transaction pool by the time the method returns.
+func (handler *TxHandler) processDecodedArray(unverifiedTxGroups []transactions.SignedTxGroup) (disconnect, allTransactionIncluded bool) {
+	var nonDuplicatedFilteredGroups bool
+	unverifiedTxGroups, nonDuplicatedFilteredGroups = handler.filterAlreadyCommitted(unverifiedTxGroups)
+
+	if len(unverifiedTxGroups) == 0 {
+		return false, !nonDuplicatedFilteredGroups
+	}
+
+	// build the transaction verification context
+	latest := handler.ledger.Latest()
+	latestHdr, err := handler.ledger.BlockHdr(latest)
+	if err != nil {
+		logging.Base().Warnf("Could not get header for previous block %v: %v", latest, err)
+		return false, false
```

Comment:
return an error to whom ? the caller doesn't have much it can do with the error.
This is a ledger internal error, that shouldn't happen. In fact, once we'll be able to remove the old transaction handler code, we could simplify it so that we get the header from the ledger only once per round.